drivers/watchdog/wdt_sifive.c

```diff
@@ -0,0 +1,289 @@
+ * @param scaler  Pointer to return scaler power of 2
+ *
+ * @return Watchdog counter value
+ */
+static int wdt_sifive_convtime(uint32_t timeout, int clk, int *scaler)
+{
+	uint64_t cnt;
+	int i;
+
+	cnt = (uint64_t)timeout * clk / 1000;
+	for (i = 0; i < 16; i++) {
+		if (cnt <= WDOGCMP_MAX) {
+			break;
+		}
+
+		cnt >>= 1;
+	}
+
+	if (i == 16) {
+		/* Maximum counter and scaler */
+		LOG_ERR("Invalid timeout value allowed range");
+
+		*scaler = WDOGCFG_SCALE_MAX;
+		return WDOGCMP_MAX;
+	}
+
+	*scaler = i;
+
+	return cnt;
+}
+
+static int wdt_sifive_install_timeout(const struct device *dev,
+				      const struct wdt_timeout_cfg *cfg)
+{
+	volatile struct wdt_sifive_reg *wdt = DEV_REG(dev);
+	struct wdt_sifive_dev_data *data = dev->data;
+	uint32_t mode = 0, t;
+	int cmp, scaler;
+
+	if (data->timeout_valid) {
+		LOG_ERR("No more timeouts can be installed");
+		return -ENOMEM;
+	}
+	if (cfg->window.min != 0U || cfg->window.max == 0U) {
+		return -EINVAL;
+	}
+
+	/*
+	 * Freedom watchdog does not support window timeout config.
+	 * So use max field of window.
+	 */
+	cmp = wdt_sifive_convtime(cfg->window.max, WDOG_CLK, &scaler);
+	if (cmp < 0 || WDOGCMP_MAX < cmp) {
+		LOG_ERR("Unsupported watchdog timeout\n");
+		return -EINVAL;
+	}
+
+	switch (cfg->flags) {
+	case WDT_FLAG_RESET_SOC:
+		/* WDT supports global SoC reset but cannot callback. */
+		mode = WDOGCFG_RSTEN | WDOGCFG_ZEROCMP;
+		break;
+	case WDT_FLAG_RESET_NONE:
+		/* No reset */
+		mode = WDOGCFG_ZEROCMP;
+		break;
+	case WDT_FLAG_RESET_CPU_CORE:
+	default:
+		LOG_ERR("Unsupported watchdog config flags\n");
+
+		wdt_sifive_disable(dev);
+		return -ENOTSUP;
+	}
+
+	t = wdt->wdogcfg;
+	t &= ~(WDOGCFG_RSTEN | WDOGCFG_ZEROCMP | WDOGCFG_SCALE_MASK);
+	t |= mode | scaler;
+
+	wdt->wdogkey = WDOG_KEY;
+	wdt->wdogcfg = t;
+	wdt->wdogkey = WDOG_KEY;
+	wdt->wdogcmp0 = cmp;
+
+	data->cb = cfg->callback;
+	data->enable_cb = true;
+	data->timeout_valid = true;
+
+	return 0;
+}
+
+static int wdt_sifive_feed(const struct device *dev, int channel_id)
+{
+	volatile struct wdt_sifive_reg *wdt = DEV_REG(dev);
+
+	wdt->wdogkey = WDOG_KEY;
+	wdt->wdogfeed = WDOG_FEED;
+
+	return 0;
+}
+
+static const struct wdt_driver_api wdt_sifive_api = {
+	.setup = wdt_sifive_setup,
+	.disable = wdt_sifive_disable,
+	.install_timeout = wdt_sifive_install_timeout,
+	.feed = wdt_sifive_feed,
+};
+
+DEVICE_DECLARE(wdt_sifive);
+
+static void wdt_sifive_irq_config(void)
+{
+	IRQ_CONNECT(DT_INST_IRQN(0),
+		    DT_INST_IRQ(0, priority), wdt_sifive_isr,
+		    DEVICE_GET(wdt_sifive), 0);
+	irq_enable(DT_INST_IRQN(0));
+}
+
+static int wdt_sifive_init(const struct device *dev)
+{
+#ifdef CONFIG_WDT_DISABLE_AT_BOOT
+	wdt_sifive_disable(dev);
+#endif
+	wdt_sifive_irq_config();
+
+	return 0;
+}
+
+static struct wdt_sifive_dev_data wdt_sifive_data;
+
+static const struct wdt_sifive_device_config wdt_sifive_cfg = {
+	.regs = DT_INST_REG_ADDR(0),
+};
+
+DEVICE_AND_API_INIT(wdt_sifive, DT_INST_LABEL(0), wdt_sifive_init,
```

Comment:
You need to update this to use DEVICE_DT_DEFINE

Reply:
Thanks for review. Done.